aptos-move/framework/aptos-framework/sources/delegation_pool.move

```diff
@@ -554,15 +790,165 @@ module aptos_framework::delegation_pool {
+            }, false, false)
+        };
+        let voting_power_delegation = *table::borrow(voting_power_delegation, delegator);
+        let needs_update = false;
+        if (voting_power_delegation.last_locked_until_secs < locked_until_secs &&
+            voting_power_delegation.voter != voting_power_delegation.voter_next_lockup) {
+            voting_power_delegation.voter = voting_power_delegation.voter_next_lockup;
+            needs_update = true;
+        };
+        (voting_power_delegation, true, needs_update)
+    }
+
+    /// Update VotingPowerDelegation of a delegator to up-to-date then borrow_mut it.
+    fun update_and_borrow_mut_delegator_voting_power_delegation(
+        pool: &DelegationPool,
+        governance_records :&mut GovernanceRecords,
+        delegator: address
+    ): &mut VotingPowerDelegation {
+        let (latest_voting_power_delegation, is_existed, needs_update) =
+            get_latest_delegator_voting_power_delegation(pool, governance_records, delegator);
+        let voting_power_delegation_table = &mut governance_records.voting_power_delegation;
+        if (!is_existed) {
+            table::add(voting_power_delegation_table, delegator, latest_voting_power_delegation);
+        };
+        let voting_power_delegation = table::borrow_mut(voting_power_delegation_table, delegator);
+        if (needs_update) {
+            *voting_power_delegation = latest_voting_power_delegation;
+        };
+        voting_power_delegation
+    }
+
+    /// Get the latest VoterVotingPower of a voter.
+    /// The 2nd bool return value is true when VoterVotingPower has already been created before calling this function.
+    /// The 3rd bool return value is true when the latest VoterVotingPower is different from the existing VotingPowerDelegation.
+    fun get_latest_voter_voting_power(
+        pool: &DelegationPool,
+        governance_records: &GovernanceRecords,
+        voter: address
+    ): (VoterVotingPower, bool, bool) {
+        let pool_address = get_pool_address(pool);
+        let locked_until_secs = stake::get_lockup_secs(pool_address);
+        let voting_power_per_voter = &governance_records.voting_power_per_voter;
+        // By default, a delegator's voter is itself.
+        if (!table::contains(voting_power_per_voter, voter)) {
+            let active_shares = get_delegator_active_shares(pool, voter);
+            let inactive_shares = get_delegator_pending_inactive_shares(pool, voter);
+            return (VoterVotingPower {
+                active_shares,
+                pending_inactive_shares: inactive_shares,
+                active_shares_next_lockup: active_shares,
+                last_locked_until_secs: locked_until_secs,
+            }, false, false)
+        };
+        let latest_voter_voting_power = *table::borrow(voting_power_per_voter, voter);
+        let needs_updated = false;
+        if (latest_voter_voting_power.last_locked_until_secs < locked_until_secs) {
```

Comment:
This doesn't take into account the voter's own voting power if they're also a delegator